command/agent/cache/lease_cache.go

```diff
@@ -0,0 +1,813 @@
+			return
+		}
+
+		c.logger.Debug("received cache-clear request", "type", req.Type, "namespace", req.Namespace, "value", req.Value)
+
+		if err := c.handleCacheClear(ctx, req.Type, req.Namespace, req.Value); err != nil {
+			// Default to 500 on error, unless the user provided an invalid type,
+			// which would then be a 400.
+			httpStatus := http.StatusInternalServerError
+			if err == errInvalidType {
+				httpStatus = http.StatusBadRequest
+			}
+			respondError(w, httpStatus, errwrap.Wrapf("failed to clear cache: {{err}}", err))
+			return
+		}
+
+		return
+	})
+}
+
+func (c *LeaseCache) handleCacheClear(ctx context.Context, clearType string, clearValues ...interface{}) error {
+	if len(clearValues) == 0 {
+		return errors.New("no value(s) provided to clear corresponding cache entries")
+	}
+
+	// The value that we want to clear, for most cases, is the last one provided.
+	clearValue, ok := clearValues[len(clearValues)-1].(string)
+	if !ok {
+		return fmt.Errorf("unable to convert %v to type string", clearValue)
+	}
+
+	switch clearType {
+	case "request_path":
+		// For this particular case, we need to ensure that there are 2 provided
+		// indexers for the proper lookup.
+		if len(clearValues) != 2 {
+			return fmt.Errorf("clearing cache by request path requires 2 indexers, got %d", len(clearValues))
+		}
+
+		// The first value provided for this case will be the namespace, but if it's
+		// an empty value we need to overwrite it with "root/" to ensure proper
+		// cache lookup.
+		if clearValues[0].(string) == "" {
+			clearValues[0] = "root/"
+		}
+
+		// Find all the cached entries which has the given request path and
+		// cancel the contexts of all the respective renewers
+		indexes, err := c.db.GetByPrefix(clearType, clearValues...)
+		if err != nil {
+			return err
+		}
+		for _, index := range indexes {
+			index.RenewCtxInfo.CancelFunc()
+		}
+
+	case "token":
+		if clearValue == "" {
+			return nil
+		}
+
+		// Get the context for the given token and cancel its context
+		index, err := c.db.Get(cachememdb.IndexNameToken, clearValue)
+		if err != nil {
+			return err
+		}
+		if index == nil {
+			return nil
+		}
+
+		c.logger.Debug("cancelling context of index attached to token")
+
+		index.RenewCtxInfo.CancelFunc()
+
+	case "token_accessor", "lease":
+		// Get the cached index and cancel the corresponding renewer context
+		index, err := c.db.Get(clearType, clearValue)
+		if err != nil {
+			return err
+		}
+		if index == nil {
+			return nil
+		}
+
+		c.logger.Debug("cancelling context of index attached to accessor")
+
+		index.RenewCtxInfo.CancelFunc()
+
+	case "all":
+		// Cancel the base context which triggers all the goroutines to
+		// stop and evict entries from cache.
+		c.logger.Debug("cancelling base context")
+		c.baseCtxInfo.CancelFunc()
+
+		// Reset the base context
+		baseCtx, baseCancel := context.WithCancel(ctx)
+		c.baseCtxInfo = &ContextInfo{
```

Comment:
This isn't thread-safe; you need to protect c.baseCtxInfo and all of its accesses with a mutex.

Reply:
Mutexing baseCtxInfo may not be required. We set the `baseCtxInfo` while creation. We usually only read it. The only place it gets modified is when we cancel it and replace it with a new one, which happens over the API call. So, I am thinking that it should be fine. LMK if you think otherwise.

Reply:
You are replacing it with a new one -- into a member variable. Which means you're modifying the member variable, and if that can happen in a request concurrent to a different request that reads it, you have a race condition. My guess is that concurrent requests are allowed in which case you definitely have a race here.

Reply:
Fixed in https://github.com/hashicorp/vault/pull/6321